keyboards/ergodox_ez/keymaps/steno/keymap.c

```diff
@@ -230,69 +200,29 @@ const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
  */
 // TxBolt over Serial
 [TXBOLT2] = LAYOUT_ergodox(
-       KC_NO,   M(NM),   M(NM),   M(NM),   M(NM),   M(NM),  KC_NO,
-       KC_NO,   M(Sl),   M(Tl),   M(Pl),   M(Hl),   M(X),   KC_NO,
-       KC_NO,   M(Sl),   M(Kl),   M(Wl),   M(Rl),   M(X),
+       KC_NO,   STN_N1,  STN_N2,  STN_N3,  STN_N4,  STN_N5,  KC_NO,
+       KC_NO,   STN_S1,  STN_TL,  STN_PL,  STN_HL,  STN_ST1, KC_NO,
+       KC_NO,   STN_S2,  STN_KL,  STN_WL,  STN_RL,  STN_ST2,
        KC_NO,   KC_NO,   KC_NO,   KC_NO,   KC_NO,   KC_NO,   KC_NO,
-       KC_NO,   KC_NO,   KC_NO,   M(Al),   M(Ol),
+       KC_NO,   KC_NO,   KC_NO,   STN_A,   STN_O,
                                            KC_NO,   KC_NO,
                                                     KC_NO,
                                   KC_NO,   KC_NO,   KC_NO,
     // right hand
-       KC_NO,    M(NM),   M(NM),   M(NM),   M(NM),   M(NM),   M(NM),
-       KC_TRNS,  M(X),    M(Fr),   M(Pr),   M(Lr),   M(Tr),   M(Dr),
-                 M(X),    M(Rr),   M(Br),   M(Gr),   M(Sr),   M(Zr),
+       KC_NO,    STN_N6,  STN_N7,  STN_N8,  STN_N9,  STN_NA,  STN_NB,
+       KC_TRNS,  STN_ST3, STN_FR,  STN_PR,  STN_LR,  STN_TR,  STN_DR,
+                 STN_ST4, STN_RR,  STN_BR,  STN_GR,  STN_SR,  STN_ZR,
        KC_NO,    KC_NO,   KC_NO,   KC_NO,   KC_NO,   KC_NO,   KC_NO,
-                          M(Er),   M(Ur),   KC_NO,   KC_NO,   KC_NO,
+                          STN_E,   STN_U,   KC_NO,   KC_NO,   KC_NO,
        KC_NO,   KC_NO,
        KC_NO,
        KC_NO,   KC_NO,   KC_NO
 ),
 };
 
-uint8_t chord[4] = {0,0,0,0};
-uint8_t pressed_count = 0;
-
-void send_chord(void)
-{
-  for(uint8_t i = 0; i < 4; i++)
-  {
-    if(chord[i])
-      virtser_send(chord[i]);
-  }
-  virtser_send(0);
-}
-
-bool process_record_user(uint16_t keycode, keyrecord_t *record)
-{
-  // We need to track keypresses in all modes, in case the user
-  // changes mode whilst pressing other keys.
-  if (record->event.pressed)
-    pressed_count++;
-  else
-    pressed_count--;
-  return true;
-}
-
-const macro_t *action_get_macro(keyrecord_t *record, uint8_t id, uint8_t opt)
-{
-  // MACRODOWN only works in this function
-
-  if (record->event.pressed) {
-    uint8_t grp = (id & GRPMASK) >> 6;
-    chord[grp] |= id;
-  }
-  else {
-    if (pressed_count == 0) {
-      send_chord();
-      chord[0] = chord[1] = chord[2] = chord[3] = 0;
-    }
-  }
-  return MACRO_NONE;
-};
-
 // Runs just one time when the keyboard initializes.
 void matrix_init_user(void) {
+    steno_set_mode(STENO_MODE_BOLT); // or STENO_MODE_GEMINI
```

Comment:
Once this is set, you shouldn't need to set it again.
This function writes to EEPROM, so it's saved.  

But this should cause no harm. 

But it may be better to add this to `eeconfig_init_user`, so that if the eeprom is reset, it's defaulted back to this value.

Reply:
OK, I didn't know that. I just discovered it does not work at all, when not called.

Should I then also update the documentation in https://github.com/qmk/qmk_firmware/blob/c0baf2a964b10d708281c704c0b049a1cf0f4914/docs/feature_stenography.md#L42-L46?

Or is it possible to add a check to https://github.com/qmk/qmk_firmware/blob/86617d8f29edc96d93459a9a6a27693debfde325/quantum/process_keycode/process_steno.c#L88-L93 to default to BOLT. I'm no C developer, therefore I don't know if 

```
  mode = eeprom_read_byte(EECONFIG_STENOMODE);
  if (!mode) {
    mode = STENO_MODE_BOLT;
  }
```

would work.

The documentation already says so: "By default QMK will speak the TX Bolt protocol but can be switched to GeminiPR; the last protocol used is stored in non-volatile memory so QMK will use the same protocol on restart."

Reply:
OK, I think it can't be `if (!mode)`, but `if (mode != STENO_MODE_BOLT && mode != STENO_MODE_GEMINI)`.

Another problem I had was that even without the change in `process_steno.c`, I could now even remove my `matrix_init_user` completely, and it still worked, because the value is now stored in the eeprom.

So I had to reset the value first. But then the `eeconfig_init_user` method is not called at all. Will this method only be called when the eeprom has never been initialized? So it can't be used to configure it to `GEMINI`, when it was previously configured to `BOLT`? Therefore having the code in `matrix_init_user` means "do it regardless of what I said before, everytime the keyboard is powered up", whereas having it in `eeconfig_init_user` means "do this if I have never ever anytime anywhen initialized the eeprom"?

In that case having the code in `matrix_init_user` seems better, unless there is a key defined to switch protocol. All other layouts that use steno also define `matrix_init_user` instead of `_eeconfig_init_user`. And we must change `process_steno.c` to fallback to `BOLT` if there is no valid value in the eeprom yet.

Reply:
Well, the enum has it listed as 0 for Bolt. 

So that would be the default value, actually.  So you wouldn't need to call anything here, unless you wanted to use Gemini. 

So really, the entire function could be removed here. 

https://github.com/qmk/qmk_firmware/blob/86617d8f29edc96d93459a9a6a27693debfde325/quantum/process_keycode/process_steno.h#L25

Reply:
That's odd, because at first I did not have that line at all, and it was not working. Maybe somehow there was something at that position in the eeprom. Maybe because I started with 2 different firmwares before switching to qmk, there was already something completely different at that position.

I will create a pull request for `process_steno.c` to check for a valid mode and if invalid default to bolt mode.

And for `ergodox_ez/steno` (and my own layout, which I may later provide for the `layouts/community/ergodox` folder) I would either opt to leave the `matrix_init_user` method, because

1. It's the save way to have the mode at the correct value, even if for some reason the value at the eeprom is at e.g. gemini mode
2. Most of the other layouts that use steno mode also do it this way
3. The documentation states exactly that

Or alternatively the layout introduces a switch to change the mode.

But this is not the layout I am using, just the layout I wanted to use as source to copy, so I just wanted to have it in good shape before doing that. I am fine with both alternatives, but since I am not the original author of the layout I wanted it to work exactly the same as before. Maybe @dragon788 as the original author of this layout has something to say about this.

Reply:
It's possible that you ad something else there, but that would be odd, too.  

Either way, I think this should be okay, as is.